src/Type/Constant/ConstantArrayType.php

```diff
@@ -669,7 +669,7 @@ public function unsetOffset(Type $offsetType): Type
 			return new self($this->keyTypes, $this->valueTypes, $this->nextAutoIndexes, $optionalKeys, false);
 		}
 
-		return new ArrayType($this->getKeyType(), $this->getItemType());
+		return new self($this->keyTypes, $this->valueTypes, $this->nextAutoIndexes, array_keys($this->keyTypes), false);
```

Comment:
This also unsets keys of types that `$offsetType` does not match. Something like this should make it even more useful:
```suggestion
		$optionalKeys = $this->optionalKeys;
		$isList = $this->isList;
		foreach ($this->keyTypes as $i => $keyType) {
			if ($offsetType->isSuperTypeOf($keyType)->yes()) {
				$optionalKeys[] = $i;
				$isList = false;
			}
		}

		$optionalKeys = array_values(array_unique($optionalKeys));
		return new self($this->keyTypes, $this->valueTypes, $this->nextAutoIndexes, $optionalKeys, $isList);
```

Reply:
great idea! @VincentLanglet make sure to also run/update the tests locally to avoid many force-pushes 😊

Reply:
Just discovered `make build`. I'll try to think about it next time.

Reply:
For the record I get the error
```
1) PHPStan\Command\ErrorFormatter\TableErrorFormatterTest::testEditorUrlWithRelativePath
Failed asserting that ' ------ ----------------\n
  [32mLine[39m   [32mFoo.php[39m\n
 ------ ----------------\n
  12     Test\n
         ✏️  rel/Foo.php\n
 ------ ----------------\n
\n
\n
[37;41m                                                                                [39;49m\n
[37;41m [ERROR] Found 1 error                                                          [39;49m\n
[37;41m                                                                                [39;49m\n
\n
' contains "editor://custom/path/rel/Foo.php".
```
on local.

Reply:
that looks like some terminal ANSI weirdness or so? how are you running the tests exactly? any special tools/terminal involved. I definitely don't get that error 🤔

Reply:
I always have the same error too running `make tests`

Reply:
I get the same kind of error when running `make` in the PhpStorm built-in terminal. `symfony/console` [disables hyperlinks](https://github.com/symfony/symfony/blob/v5.4.13/src/Symfony/Component/Console/Formatter/OutputFormatterStyle.php#L98-L99) because it appears to not handle them correctly.

I just run `make` in a regular terminal instead, or I ignore the error.

Reply:
I run it inside phpstorm too